signer/src/stacks/contracts.rs

```diff
@@ -77,14 +93,43 @@ pub trait AsContractCall {
     /// post-conditions, which is the address that sent the asset. The
     /// default is that we do not enforce any conditions since we usually
     /// deployed the contract.
-    fn post_conditions(&self, _: StacksAddress) -> StacksTxPostConditions {
+    fn post_conditions(&self) -> StacksTxPostConditions {
         StacksTxPostConditions {
             post_condition_mode: TransactionPostConditionMode::Allow,
             post_conditions: Vec::new(),
         }
     }
 }
 
+/// A generic newtype that implements AsTxPayload for all types that
+/// implement AsContractCall.
+///
+/// # Notes
+///
+/// Ideally, every type that implements AsContractCall should implement
+/// AsTxPayload automatically. What we want is to have something like the
+/// following:
+///
+/// impl<T: AsContractCall> AsTxPayload for T { ... }
+///
+/// But that would preclude us from adding something like:
+///
+/// impl<T: AsSmartContract> AsTxPayload for T { ... }
+///
+/// since doing so is prevented by the compiler because it introduces
+/// ambiguity. One work-around is to use a wrapper type that implements the
+/// trait that we want.
+pub struct ContractCall<T: AsContractCall>(pub T);
```

Comment:
Nit: While it's tempting to put trait bounds on structs it rarely provides any benefits. I think [this stack overflow discussion](https://stackoverflow.com/questions/49229332/should-trait-bounds-be-duplicated-in-struct-and-impl) explains it better than I can if you want to dig deeper, but the short story is that it is redundant and may add undesired limitations for future extensions.

```suggestion
pub struct ContractCall<T>(pub T);
```